File: LinkedList/java/Problem086_PartitionList.java
/**
 * Problem 86: Partition List
 * Difficulty: Medium
 * 
 * Given a linked list and a value x, partition it such that all nodes less than x 
 * come before nodes greater than or equal to x.
 * 
 * Example: [1,4,3,2,5,2], x = 3 → [1,2,2,4,3,5]
 * Time: O(n), Space: O(1)
 */
class ListNode {
    int val;
    ListNode next;
    ListNode(int val) { this.val = val; }
}

public class Problem086_PartitionList {
    public ListNode partition(ListNode head, int x) {
        ListNode lessHead = new ListNode(0), greaterHead = new ListNode(0);
        ListNode less = lessHead, greater = greaterHead;
        
        while (head != null) {
            if (head.val < x) {
                less.next = head;
                less = less.next;
            } else {
                greater.next = head;
                greater = greater.next;
            }
            head = head.next;
        }
        greater.next = null;
        less.next = greaterHead.next;
        return lessHead.next;
    }
}

File: LinkedList/python/problem_086_partition_list.py
"""Problem 86: Partition List - Medium
Given a linked list and value x, partition nodes < x before nodes >= x
Example: [1,4,3,2,5,2], x=3 → [1,2,2,4,3,5]
Time: O(n), Space: O(1)
"""
class ListNode:
    def __init__(self, val=0, next=None):
        self.val, self.next = val, next

class Solution:
    def partition(self, head: ListNode, x: int) -> ListNode:
        less_head, greater_head = ListNode(0), ListNode(0)
        less, greater = less_head, greater_head
        
        while head:
            if head.val < x:
                less.next = head
                less = less.next
            else:
                greater.next = head
                greater = greater.next
            head = head.next
        greater.next = None
        less.next = greater_head.next
        return less_head.next
